### ISSUE TITLE
Miri does not catch UB with manual Rc::get_mut when another Rc exists

### ISSUE THREAD

**Issue by hkBst** [rust-lang/rust#136126]
I tried this code:

```rust
use std::rc::Rc;

fn main() {
    let mut foo = Rc::new(1u8);

    {
        let _foo2 = foo.clone();
        assert_eq!(Rc::get_mut(&mut foo), None);
        // get_mut-like unsafe code, UB because of foo2?
        let foo_ref = unsafe { &mut *(Rc::as_ptr(&foo) as *mut u8) };
        *foo_ref += 1;
    }

    assert_eq!(Rc::into_inner(foo), Some(2));
}
```

I expected Miri to complain about UB, but it did not.

Using Rust 1.86.0-nightly (2025-01-26 f85c6de55206dbee5ffe)

**Comment by meithecatte** [comment]
My first intuition was that this is the case because you aren't touching `foo2` afterwards, and so you don't actually trigger the UB. It seems that I'm wrong here though, because even if you add `assert_eq!(*foo2, 2);`, miri still doesn't complain.
